### PR TITLE
[WIP]add cpu version batch normalization

### PR DESCRIPTION
**Pull request by sunalbert**
As mentioned in #122，dynet does not support batch normalization and layer normalization operation. I am trying to add batch normalization(https://arxiv.org/pdf/1502.03167.pdf), which is useful in GAN and CNN. The task has some sub-tasks:
- [ ] batch statistics operation: to get mean and variance of batch data. This operation also needs compute the moving_average and moving_variance which are used in test mode.I find there is a similar PR #385, maybe I can implement batch statistics with it.
- [x] batch normalization operation: based on the mean and variance provided by batch statistics, normalize the batch data.
- [ ] necessary comments and documents for batch statistics and batch normalization.
- [ ] unit test and some detailed debugging

### CODE REVIEW

**Comment by sunalbert** [comment]
I find a Average operation in dynet, but it is not applicable for batch normalization, because batch normalization is a channel-wise operation and needs compute moving average. I'd like to add a new batch_mean operation rather than modify Average operation directly so that we can avoid compatibility problems. What do you think about it? any better idea? @neubig

**Comment by pmichel31415** [comment]
Hi, just chiming in to share an idea I had when I though of implementing batchnorm in dynet a while back : 

would it be easier to implement it as a separate class (much like RNNBuilder or SoftmaxBuilder, something like NormalizationLayer...)? So you can store the statistics as attributes and you don't need to change the expression interface to accommodate a train/test switch.

Also this could be extended to implement BatchReNorm, LayerNorm, etc...

With this the only actual nodes that need to be implemented are the averaging, broadcasted multiplication/division, etc..

**Comment by sunalbert** [comment]
Hi @pmichel31415, thanks for your reply. In my original idea, I should implement average operation, variance operation, batchnorm operation. They can be called as following:
`mean = mean(input, train_or_test)`
`variance = variance(input, train_or_test)`
`out = bacthnorm(mean, variance)`
If we implement batch norm as a  separate class, user can call batch normalization operation in one line code:
`out =  batchnorm(input)`
Right?
